pkg/queues/amqp_provider.go

```diff
@@ -0,0 +1,189 @@
+package queues
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/flightctl/flightctl/pkg/log"
+	"github.com/flightctl/flightctl/pkg/reqid"
+	"github.com/go-chi/chi/v5/middleware"
+	amqp "github.com/rabbitmq/amqp091-go"
+	"github.com/sirupsen/logrus"
+)
+
+type amqpProvider struct {
+	ctx     context.Context
+	cancel  context.CancelFunc
+	url     string
+	log     logrus.FieldLogger
+	wg      *sync.WaitGroup
+	queues  []*amqpQueue
+	stopped bool
+}
+
+func NewAmqpProvider(url string, log logrus.FieldLogger) Provider {
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(context.Background())
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		<-ctx.Done()
+	}()
+	return &amqpProvider{
+		ctx:    ctx,
```

Comment:
question: do we need to embed ctx vs pass as a param?

Reply:
When the provider is stopped the ctx is canceled.  This should cause the consume loop to stop.  If we pass it as param, the consume loop will not be stopped.

Reply:
I think the general problem is that context is meant to be passed as a param not internalized in a struct[1].  You can achieve the same thing and only register the cancelFn to the struct.

[1] https://go.dev/blog/context-and-structs

Reply:
How is it the same?  The cancelFn is used to cancel (close) the `ctx.Done()` channel.  The ctx is used to get indication that channel was closed.

Reply:
OK.  Maybe I can avoid using context.  Instead I can pass a channel that will be closed when the provider is stopped.  Do you think it is better ?

Reply:
Context is useful if you expect the underlying implementation to take a context or if you want to use it for go routine synchronization.  But you want to keep the context external. This assists in graceful termination. But could you just `close(doneCh)`. That is your choice.

```go
// higher level context
ctx, cancelFn :=  context.WithCancel(context.Background())

// give the cancelFn to some actor who controls the lifecycle of the provider?

provider, err  := NewProvider()
// handle err

// here my Run loop which takes a context inside Run I create a child context and pin that cancelFn to the struct.
go provider.Run(ctx)

<-ctx.Done() // Wait?

```

Reply:
OK